alicloud/resource_alicloud_router_interface.go

```diff
@@ -0,0 +1,297 @@
+	if err != nil {
+		return fmt.Errorf("DescribeRouterInterfaces got an error: %#v", err)
+	}
+
+	routerInterface := resp.RouterInterfaceSet.RouterInterfaceType
+	if len(routerInterface) == 0 {
+		return fmt.Errorf("No router interface found.")
+	}
+
+	for _, ri := range routerInterface {
+		if ri.RouterInterfaceId == d.Id() {
+			d.Set("role", ri.Role)
+			d.Set("spec", ri.Spec)
+			d.Set("name", ri.Name)
+			d.Set("router_id", ri.RouterId)
+			d.Set("router_type", ri.RouterType)
+			d.Set("description", ri.Description)
+			d.Set("access_point_id", ri.AccessPointId)
+			d.Set("opposite_access_point_id", ri.OppositeAccessPointId)
+			d.Set("opposite_router_type", ri.OppositeRouterType)
+			d.Set("opposite_router_id", ri.OppositeRouterId)
+			d.Set("opposite_interface_id", ri.OppositeInterfaceId)
+			d.Set("opposite_interface_owner_id", ri.OppositeInterfaceOwnerId)
+			d.Set("health_check_source_ip", ri.HealthCheckSourceIp)
+			d.Set("health_check_target_ip", ri.HealthCheckTargetIp)
+			return nil
+		}
+	}
+
+	d.SetId("")
+	return fmt.Errorf("No router interface found.")
+}
+
+func resourceAlicloudRouterInterfaceDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AliyunClient).ecsconn
+
+	args := &ecs.OperateRouterInterfaceArgs{
+		RegionId:          getRegion(d, meta),
+		RouterInterfaceId: d.Id(),
+	}
+
+	return resource.Retry(5*time.Minute, func() *resource.RetryError {
+		if _, err := conn.DeleteRouterInterface(args); err != nil {
+			if IsExceptedError(err, IncorrectStatus) || IsExceptedError(err, DependencyViolationRouterInterfaceReferedByRouteEntry) {
+				return resource.RetryableError(fmt.Errorf("Router interface in use - trying again while it is deleted."))
+			}
+			return resource.NonRetryableError(fmt.Errorf("Error deleting interface %s: %#v", d.Id(), err))
+		}
+		return nil
+	})
+}
+
+func buildAlicloudRouterInterfaceCreateArgs(d *schema.ResourceData, meta interface{}) (*ecs.CreateRouterInterfaceArgs, error) {
+	args := &ecs.CreateRouterInterfaceArgs{
+		RegionId:         getRegion(d, meta),
+		RouterType:       ecs.RouterType(d.Get("router_type").(string)),
+		RouterId:         d.Get("router_id").(string),
+		Role:             ecs.Role(d.Get("role").(string)),
+		Spec:             ecs.Spec(d.Get("spec").(string)),
+		OppositeRegionId: getRegion(d, meta),
+	}
+
+	if v, ok := d.GetOk("opposite_router_type"); ok {
+		args.OppositeRouterType = ecs.RouterType(v.(string))
+	}
+
+	if args.RouterType == ecs.VBR {
+		if args.Role != ecs.InitiatingSide {
+			return nil, fmt.Errorf("The 'role' must be 'InitiatingSide' when the 'router_type' value is 'VBR'.")
+		}
+
+		if args.OppositeRouterType != ecs.VRouter {
+			return nil, fmt.Errorf("The 'role' must be 'InitiatingSide' when the 'router_type' value is 'VBR'.")
+		}
+
+		v, ok := d.GetOk("access_point_id")
+		if !ok {
+			return nil, fmt.Errorf("The 'access_point_id' must be set when the 'router_type' value is 'VBR'.")
```

Comment:
please modify error log: "'access_point_id': required field is not set when 'router_type' is 'VBR'."